### PR TITLE
refine code

### PR DESCRIPTION
**Pull request by cwen0**
* use storage pkg to store jobs/model template/model instance  
* add schedule queue to schedule jobs 
* Jobs/models add state control

### CODE REVIEW

File: oasis/app.py [hunk 1]
```diff
@@ -49,7 +51,11 @@ def parse_config():
         logger.error("models path is required!!")
         sys.exit(1)
 
-    model.MODELS_PATH = options.models_path
+    if options.db_path == "":
```
Comment (zhouqiang-cl) [inline oasis/app.py]: set default path

File: oasis/app.py [hunk 2]
```diff
@@ -71,7 +77,7 @@ def stop_loop(deadline):
     def shutdown():
         logger.info('Stopping http server')
         server.stop()
-        JobHandler.close_all_jobs()
+        handler.close()
```
Comment (zhouqiang-cl) [inline oasis/app.py]: close -> stop?

File: oasis/controller.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    def schedule_task(self):
+        if self.queue.empty():
+            logger.info("the queue of tasks is empty, skip schedule")
+            return
+
+        if len(self.jobs) >= RUNNING_SIZE_LIMIT:
+            logger.warn("the number of tasks allowed to run exceeded the limit.")
+            return
+
+        task = self.queue.get()
```
Comment (zhouqiang-cl) [inline oasis/controller.py]: use `get_nowait`  and remove line 73-75, check task if it is None here

File: oasis/controller.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+        t = Thread(target=job_instance.close)
+        t.start()
+
+    def clean(self):
+        delete_ids = []
+
+        for job_id, job in self.jobs.items():
+            status = job.status
+            if status == JOB_ERROR or status == JOB_STOPPED or status == JOB_FINISHED:
```
Comment (zhouqiang-cl) [inline oasis/controller.py]: can define a list 
```
clean_list = [JOB_ERROR, JOB_STOPPED, JOB_FINISHED]
if status in clean_list:
       ...
```

File: oasis/handler.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+from oasis.libs.log import logger
+from oasis.job import DEFAULT_JOB_TIMEOUT
+from oasis.libs.iexceptions import (
+    NewJobException,
+    JobNotExistsException,
+    JobNotRunningException
+)
+
+HTTP_MISS_ARGS = 401
+HTTP_ARGS_NOT_VALID = 402
```
Comment (zhouqiang-cl) [inline oasis/handler.py]: where to use this HTTP code

File: oasis/job.py [hunk 6]
```diff
@@ -3,102 +3,183 @@
...
 DEFAULT_JOB_TIMEOUT = "240h"
 THREAD_JOIN_TIMEOUT = 5
 REPORT_ADDRESS = ""
+SYNC_INTERVAL = "1m"
+
+JOB_PENDING = 'pending'
+JOB_RUNNING = "running"
+JOB_ERROR = "error"
+JOB_FINISHED = "finished"
+JOB_STOPPED = "stopped"
```
Comment (zhouqiang-cl) [inline oasis/job.py]: did we need job init. etc

File: oasis/job.py [hunk 7]
```diff
@@ -3,102 +3,183 @@
...
-    def close(self):
-        logger.info("[job-id:{job_id}] start to stop"
+    def run(self):
+        with self.lock:
+            logger.info("[job:{job}] start to run"
+                        .format(job=self.to_dict()))
+            try:
+                for md in self.api_models_config:
+                    model = Models[md.get("name")]
+                    md_store = self.store.set_model_instance({
```
Comment (zhouqiang-cl) [inline oasis/job.py]: rename ` md_store -> md_instance`?

File: oasis/controller.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+                delete_ids.append(job_id)
+
+        for job_id in delete_ids:
+            del self.jobs[job_id]
+
+
+START_JOB = "start job"
```
Comment (zhouqiang-cl) [inline oasis/controller.py]: `START_JOB = "start"`

File: oasis/handler.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+
+class JobNewHandler(tornado.web.RequestHandler):
+    def post(self):
+        try:
+            data = json.loads(self.request.body)
+        except Exception as e:
+            logger.error("load json: {json_data}\nfailed: {err}"
+                         .format(json_data=self.request.body, err=str(e)))
+            logger.exception("Exception Logged")
```
[the remark this answers is not in the document]
Comment (cwen0) [reply oasis/handler.py]: print "errro stack“， just for debug

File: oasis/app.py (hunk 2 above)
Comment (cwen0) [reply oasis/app.py]: yep, `handler.close()` use `manager` to close.

File: oasis/job.py (hunk 6 above)
Comment (cwen0) [reply oasis/job.py]: these "status' is enough for now

**Comment by cwen0** [comment]
@zhouqiang-cl PTAL, again

**Pull request merged**